lib/clientv2/watch_e2e_test.go

```diff
@@ -0,0 +1,182 @@
+						ASNumber: numorstring.ASNumber(ii),
+					},
+				}
+				_, outError := c.BGPPeers().Create(ctx, peer, options.SetOptions{})
+				Expect(outError).NotTo(HaveOccurred())
+				deleteEvents = append(deleteEvents, watch.Event{
+					Type:     watch.Deleted,
+					Previous: peer,
+				})
+			}
+
+			By("Listing all the resources")
+			outList, outError := c.BGPPeers().List(ctx, options.ListOptions{})
+			Expect(outError).NotTo(HaveOccurred())
+			Expect(outList.Items).To(HaveLen(numEvents))
+
+			By("Creating a watcher, watching the current resource version")
+			w, _ := c.BGPPeers().Watch(ctx, options.ListOptions{ResourceVersion: outList.ResourceVersion})
+			testWatcher := testutils.TestResourceWatch(w)
+			defer testWatcher.Stop()
+
+			By("Cleaning the datastore and expecting deletion events for each configured resource (tests prefix deletes results in individual events for each key)")
+			be.Clean()
+			testWatcher.ExpectEvents(apiv2.KindBGPPeer, deleteEvents)
+
+			By("Stopping the watcher and checking for full termination")
+			testWatcher.Stop()
+			Eventually(w.(*watcher).hasTerminated).Should(BeTrue())
+		})
+	})
+
+	Describe("Test constant stream of events whilst closing watcher", func() {
+		It("should handle gracefully closing watchers while events are occurring", func() {
+			c, err := New(config)
+			Expect(err).NotTo(HaveOccurred())
+
+			be, err := backend.NewClient(config)
+			Expect(err).NotTo(HaveOccurred())
+			be.Clean()
+
+			// We create a number of threads in this test.  Use a context as a simple mechanism
+			// for cancelling those threads.
+			finishctx, finish := context.WithCancel(ctx)
+
+			By("Creating an asynchronous events generator")
+			wg := sync.WaitGroup{}
+
+			// Create a goroutine to generate add/delete events.
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for {
+					if _, err := c.BGPPeers().Create(ctx, &apiv2.BGPPeer{
+						ObjectMeta: metav1.ObjectMeta{Name: "name1"},
+						Spec: apiv2.BGPPeerSpec{
+							PeerIP:   "1.2.3.4",
+							ASNumber: numorstring.ASNumber(12345),
+						},
+					}, options.SetOptions{}); err != nil {
+						panic(err)
+					}
+					time.Sleep(5 * time.Millisecond)
+					if err := c.BGPPeers().Delete(ctx, "name1", options.DeleteOptions{}); err != nil {
+						panic(err)
+					}
+					time.Sleep(5 * time.Millisecond)
+
+					if finishctx.Err() != nil {
+						break
+					}
+				}
+
+			}()
+
+			// Create a number of goroutines to handle a Watcher.  Each go routine will pull
+			// events off the watcher and exit once the results channel is closed.
+			watchers := make([]watch.Interface, numWatchers)
+			for i := 0; i < numWatchers; i++ {
+				watchers[i], err = c.BGPPeers().Watch(ctx, options.ListOptions{})
+				Expect(err).NotTo(HaveOccurred())
+				defer watchers[i].Stop()
+
+				wg.Add(1)
+				go func(w watch.Interface) {
+					defer wg.Done()
+					for e := range w.ResultChan() {
+						if e.Type == watch.Deleted {
+							if e.Previous == nil {
```

Comment:
Since this is a test a better thing to write would be `Expect(e.Previous).NotTo(BeNil(), "no previous value on deleted event")`, same for the following assertions

Reply:
Hmm, I'll try, but I had a feeling that wouldn't work within a goroutine.

Reply:
Oh, it seems to work.  I'll update the ones above too then.